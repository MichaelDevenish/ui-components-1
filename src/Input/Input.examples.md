# Input

## Basic Input

```jsx
  <div>
    Text
    <TextInput />
    Number
    <NumberInput />
    Checkbox
    <CheckboxInput 
      name='example-checkbox'
      label='This is a checkbox input :)'
    />
  </div>
```

## Input with Attributes

```jsx
  <div>
    Text
    <TextInput
     maxlength={3}
    />
    Number
    <NumberInput 
      min={0}
      max={5}
      step={0.5}
      placeholder={2.5}
    />
  </div>
```

## Input with Icon

```jsx
  <div>
    Text
    <TextInput 
      icon={<FontAwesomeIcon type='facebook' />}
    />
    Number
    <NumberInput 
      icon={<IntelliIcon type='search' />}
    />
  </div>
```

## Input with highlightOnFocus

```jsx
   <div>
     Text
     <TextInput 
       value='Blah blah I am text'
       highlightOnFocus
     />
     Number
     <NumberInput 
       value={3}
       highlightOnFocus
     />
   </div>
```

## Input with handleBlur

```jsx
   <div>
     Text
     <TextInput 
       value='Blur me!'
       handleBlur={(e, value) => alert(`
            event: ${e}
            value: ${value}
       `)}
     />
     Number
     <NumberInput 
       value={987654321}
       handleBlur={(e, value) => alert(`
           event: ${e}
           value: ${value}
       `)}
     />
   </div>
```

## Invalid Input

```jsx
  <TextInput 
    isInvalid
  />
```

## Disabled Input

```jsx
  <TextInput
    isDisabled
   />
```
