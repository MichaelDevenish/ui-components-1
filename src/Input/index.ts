export { NumberInput } from './NumberInput'
export { TextInput } from './TextInput'
export { CheckboxInput } from './CheckboxInput'
