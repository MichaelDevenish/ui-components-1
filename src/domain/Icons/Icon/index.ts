export { Icon, IIconProps } from './Icon'
export {
  IconType,
  FontAwesomeIconType,
  FontAwesomeIconTypeNoPrefix,
  IntelliIconType,
  IntelliIconTypeNoPrefix
} from './types'
