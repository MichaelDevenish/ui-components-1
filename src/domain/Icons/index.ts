export * from './FontAwesomeIcon'
export * from './IntelliIcon'
export * from './Icon'
