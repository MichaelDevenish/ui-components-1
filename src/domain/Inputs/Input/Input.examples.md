#### Basic Input

```jsx
const { TextInput } = require('../TextInput');
const { NumberInput } = require('../NumberInput');
const { CheckboxInput } = require('../CheckboxInput');

initialState = { value: false };

<div>
  Text
  <TextInput />
  Number
  <NumberInput />
  Checkbox
  <CheckboxInput
    name='example-checkbox'
    label='This is a checkbox input :)'
    value={state.value}
    handleChange={() => setState({value: !state.value})}
  />
</div>
```

#### Input with Attributes

```jsx
const { TextInput } = require('../TextInput');
const { NumberInput } = require('../NumberInput');

<div>
  Text
  <TextInput
   maxlength={3}
  />
  Number
  <NumberInput
    min={0}
    max={5}
    step={0.5}
    placeholder={2.5}
  />
</div>
```

#### Input with Icon

```jsx
const { TextInput } = require('../TextInput');
const { NumberInput } = require('../NumberInput');
const { FontAwesomeIcon } = require('@Domain/Icons');
const { IntelliIcon } = require('@Domain/Icons');

<div>
  Text
  <TextInput
    icon={<FontAwesomeIcon type='facebook' />}
  />
  Number
  <NumberInput
    icon={<IntelliIcon type='search' />}
  />
</div>
```

#### Input with Disable Prefix text

```jsx
const { TextInput } = require('../TextInput');
const { NumberInput } = require('../NumberInput');
const { FontAwesomeIcon } = require('@Domain/Icons');

<div>
  Text
  <TextInput
    icon={<FontAwesomeIcon type='facebook' />}
    disabledPrefix='Hello!' 
  />
  Number
  <NumberInput
    disabledPrefix='(123)' 
  />
</div>
```

#### Input with highlightOnFocus

```jsx
const { TextInput } = require('../TextInput');
const { NumberInput } = require('../NumberInput');

<div>
  Text
  <TextInput
    value='Blah blah I am text'
    highlightOnFocus
  />
  Number
  <NumberInput
    value={3}
    highlightOnFocus
  />
</div>
```

#### Input with handleChange

```jsx
const { TextInput } = require('../TextInput');
const { NumberInput } = require('../NumberInput');

<div>
  Text
  <TextInput
    value='Blur me!'
    handleChange={(e) => alert(`
      event: ${e}
      value: ${e.target.value}
    `)}
  />
  Number
  <NumberInput
    value={987654321}
    handleChange={(e) => alert(`
      event: ${e}
      value: ${e.target.value}
    `)}
  />
</div>
```

#### Invalid Input

```jsx
const { TextInput } = require('../TextInput');

<TextInput
  isInvalid
/>
```

#### Disabled Input

```jsx
const { TextInput } = require('../TextInput');

<TextInput
  isDisabled
/>
```
