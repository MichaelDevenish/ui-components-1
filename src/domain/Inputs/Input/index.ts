export {
  GenericInputProps,
  InputProps,
  Input
} from './Input'
