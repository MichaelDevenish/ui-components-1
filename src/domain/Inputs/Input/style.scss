:local(.input) {
  &[type='text'], &[type='password'], &[type='date'], &[type='email'], &[type='number'], &[type='url'] {
    margin: 0;
  }

  &.input-group-left {
    border-radius: 4px 0 0 4px;
  }

  &.input-group-middle {
    border-radius: 0;
  }

  &.input-group-right {
    border-radius: 0 4px 4px 0;
  }
}
