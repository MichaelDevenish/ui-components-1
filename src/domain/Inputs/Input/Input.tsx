import classNames from 'classnames'
import { get, isNil } from 'lodash'
import React, { ChangeEventHandler, RefObject } from 'react'

import { Props } from '../../../common'
import { DisabledTextWrapper, InputWrapper, PrefixWrapper, StyledInput } from '../services/style'
import { InputGroupPosition } from '../InputGroup'

const style = require('./style.scss')

interface IGenericInputProps {
  /** ID of the input */
  id?: string
  /** Name of the input */
  name: string
  /** Custom classname to use */
  className?: string
  /** If true, adds invalid input class to component */
  isInvalid?: boolean
  /** Function passed to `onChange` prop */
  handleChange?: ChangeEventHandler<HTMLInputElement>
  /** Value of the input */
  value?: string | number
  /** [Number only] Minimum value allowed */
  min?: number
  /** [Number only] Maximum value allowed */
  max?: number
  /** If true, sets input to disabled state */
  isDisabled?: boolean
  /** Icon to display in the input box */
  icon?: JSX.Element
  /** Highlight input value on focus */
  highlightOnFocus?: boolean
  /** Handle blur event */
  handleBlur?: (e: React.FocusEvent<HTMLInputElement>, value?: string | number) => void
  /** Handle key down events */
  handleKeyDown?: (e: React.KeyboardEvent<HTMLInputElement>) => void
  /** Max length of string for text input */
  maxLength?: number
  /** Placeholder text to display when input is empty */
  placeholder?: string
  /** If true, use HTML5 required attribute */
  isHTML5Required?: boolean
  /** Add autofocus attribute to input */
  autoFocus?: boolean
  /** Specify if input is checked */
  isChecked?: boolean
  /** Specify where a input is in an InputGroup */
  groupPosition?: InputGroupPosition
  /** Specify width of the input */
  width?: string
  /** Disable Prefix to display in the input box */
  disabledPrefix?: string
}

interface InputProps extends IGenericInputProps {
  /** Type of input to display */
  type: string
}

export class Input extends React.PureComponent<InputProps> {
  get classNames (): string {
    const {
      className,
      isInvalid,
      groupPosition
    } = this.props

    return classNames(
      style.input,
      [
        className,
        {
          'is-invalid-input': isInvalid,
          [`input-group-${groupPosition}`]: !isNil(groupPosition)
        }
      ]
    )
  }

  private prefixWrapperRef: RefObject<HTMLDivElement> = React.createRef()

  public componentDidMount () {
    const {
      icon,
      disabledPrefix
    } = this.props

    if (icon || disabledPrefix) {
      // forceUpdate is required to for the ref to work
      this.forceUpdate()
    }
  }

  public componentDidUpdate (prevProps: InputProps) {
    const {
      icon,
      disabledPrefix
    } = this.props

    if (
      (icon || disabledPrefix) &&
      (!prevProps.icon && !prevProps.disabledPrefix) ||
      disabledPrefix !== prevProps.disabledPrefix
    ) {
      // forceUpdate is required to for the ref to work
      this.forceUpdate()
    }
  }

  public onFocus = (e: React.FocusEvent<HTMLInputElement>): void => {
    const {
      highlightOnFocus
    } = this.props

    if (highlightOnFocus) {
      const input = e.target as HTMLInputElement
      input.select()
    }
  }

  public input (): JSX.Element {
    const {
      id,
      name,
      type,
      handleChange,
      isDisabled,
      value,
      handleBlur,
      min,
      max,
      maxLength,
      placeholder,
      isHTML5Required,
      autoFocus,
      handleKeyDown,
      isChecked,
      width
    } = this.props

    return (
      <StyledInput
        id={id || name}
        name={name}
        type={type}
        value={value}
        checked={isChecked}
        onChange={handleChange}
        onKeyDown={handleKeyDown}
        onBlur={handleBlur ? (e) => handleBlur(e, value) : undefined}
        onFocus={this.onFocus}
        className={this.classNames}
        disabled={isDisabled}
        placeholder={placeholder}
        min={min}
        max={max}
        maxLength={maxLength}
        required={isHTML5Required}
        autoFocus={autoFocus}
        style={width ? { width } : undefined}
        data-component-type={Props.ComponentType.TextInput}
      />
    )
  }

  public render (): JSX.Element {
    const {
      icon,
      disabledPrefix
    } = this.props

    if (icon || disabledPrefix) {
      return (
        <InputWrapper
          disabledPrefix={disabledPrefix}
          hasIcon={!!icon}
          prefixWrapperWidth={get(this.prefixWrapperRef, 'current.clientWidth', 0)}
        >
          <PrefixWrapper
            innerRef={this.prefixWrapperRef}
          >
            {icon}
            <DisabledTextWrapper> {disabledPrefix} </DisabledTextWrapper>
          </PrefixWrapper>
          {this.input()}
        </InputWrapper>
      )
    }

    return this.input()
  }
}

export {
  IGenericInputProps,
  InputProps
}
