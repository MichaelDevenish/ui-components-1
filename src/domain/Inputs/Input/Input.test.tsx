import { shallow } from 'enzyme'
import React from 'react'

import { Input } from './Input'
import { Icon } from '@Domain/Icons'

describe('<Input />', () => {
  it(`should render an input`, () => {
    const wrapper = shallow(
      <Input
        name='test-input'
        type='text'
      />
    )

    expect(wrapper).toMatchSnapshot()
  })

  it(`should render an input with an icon`, () => {
    const wrapper = shallow(
      <Input
        name='test-input'
        type='text'
        icon={
          <Icon
            type='fa-check'
          />
        }
      />
    )

    expect(wrapper).toMatchSnapshot()
  })

  it(`should render with an onBlur function if handleBlur is passed in`, () => {
    const mockHandleBlur = jest.fn()

    const wrapper = shallow(
      <Input
        name='test-input'
        type='text'
        handleBlur={mockHandleBlur}
      />
    )

    wrapper.simulate('blur')

    expect(mockHandleBlur.mock.calls.length).toBe(1)
    expect(wrapper).toMatchSnapshot()
  })

  it(`should highlight all text when focused and highlightOnFocus is true`, () => {
    const mockedEvent = {
      target: {
        select: jest.fn()
      }
    }

    const wrapper = shallow(
      <Input
        name='test-input'
        type='text'
        highlightOnFocus={true}
      />
    )

    wrapper.simulate('focus', mockedEvent)

    expect(mockedEvent.target.select.mock.calls.length).toBe(1)
    expect(wrapper).toMatchSnapshot()
  })

  it(`should not highlight all text when focused and highlightOnFocus is false\``, () => {
    const mockedEvent = {
      target: {
        select: jest.fn()
      }
    }

    const wrapper = shallow(
      <Input
        name='test-input'
        type='text'
        highlightOnFocus={false}
      />
    )

    wrapper.simulate('focus', mockedEvent)

    expect(mockedEvent.target.select.mock.calls.length).toBe(0)
    expect(wrapper).toMatchSnapshot()
  })
})
