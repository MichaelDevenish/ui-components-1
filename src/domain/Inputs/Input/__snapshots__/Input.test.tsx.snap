// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Input /> should highlight all text when focused and highlightOnFocus is true 1`] = `
<styled.input
  className="input "
  id="test-input"
  name="test-input"
  onFocus={[Function]}
  type="text"
/>
`;

exports[`<Input /> should not highlight all text when focused and highlightOnFocus is false\` 1`] = `
<styled.input
  className="input "
  id="test-input"
  name="test-input"
  onFocus={[Function]}
  type="text"
/>
`;

exports[`<Input /> should render an input 1`] = `
<styled.input
  className="input "
  id="test-input"
  name="test-input"
  onFocus={[Function]}
  type="text"
/>
`;

exports[`<Input /> should render an input with a disable prefix text 1`] = `
<styled.div
  disabledPrefix="test"
  hasIcon={false}
  prefixWrapperWidth={0}
>
  <styled.div
    innerRef={
      Object {
        "current": null,
      }
    }
  >
    <styled.span>
       
      test
       
    </styled.span>
  </styled.div>
  <styled.input
    className="input "
    id="test-input"
    name="test-input"
    onFocus={[Function]}
    type="text"
  />
</styled.div>
`;

exports[`<Input /> should render an input with an icon 1`] = `
<styled.div
  hasIcon={true}
  prefixWrapperWidth={0}
>
  <styled.div
    innerRef={
      Object {
        "current": null,
      }
    }
  >
    <Icon
      isSpinning={false}
      size="medium"
      type="fa-check"
    />
    <styled.span>
       
       
    </styled.span>
  </styled.div>
  <styled.input
    className="input "
    id="test-input"
    name="test-input"
    onFocus={[Function]}
    type="text"
  />
</styled.div>
`;

exports[`<Input /> should render with an onBlur function if handleBlur is passed in 1`] = `
<styled.input
  className="input "
  id="test-input"
  name="test-input"
  onBlur={[Function]}
  onFocus={[Function]}
  type="text"
/>
`;
